src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,22 +1,80 @@
 package com.es.phoneshop.model.product;
 
 import java.util.ArrayList;
+import java.util.Comparator;
 import java.util.List;
 import java.util.Optional;
 import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
-    private List<Product> productsList = new ArrayList<>();
+
+    private static volatile ArrayListProductDao instance;
+    private List<Product> productsList;
+
+    private ArrayListProductDao() {
+        productsList = new ArrayList<>();
+    }
+
+    public static ArrayListProductDao getInstance() {
+        ArrayListProductDao tempInstance = instance;
+        if (tempInstance == null) {
+            synchronized (ArrayListProductDao.class) {
+                tempInstance = instance;
+                if (tempInstance == null) {
+                    instance = tempInstance = new ArrayListProductDao();
+                }
+            }
+        }
+
+        return tempInstance;
+    }
 
     @Override
-    public synchronized Product getProduct(Long id) {
+    public synchronized Product getProduct(final Long id) {
         return productsList.stream().filter(p -> p.getId().equals(id)).findFirst().orElse(null);
     }
 
     @Override
-    public synchronized List<Product> findProducts() {
-        return productsList.stream().filter(p -> (p.getPrice() != null) && (p.getStock() > 0)).
-                collect(Collectors.toList());
+    public synchronized List<Product> findProducts(final String query, final String sortField, final String sortMode) {
+        return productsList.stream().filter(p -> (p.getPrice() != null) && (p.getStock() > 0))
+                .filter(p -> {
```

Comment:
make it more readable :)
hints:
comparator = comparator.thenComparing(comparator2).reversed();
filter = predicat.and(predicat2);
productsList.stream().filter(filter).sorted(comparator).collect(toList());